### PR TITLE
ARC-1414 Add isAdmin check at the start of every resolver func

### PR DESCRIPTION
**Pull request by yvonnevictoria**
Add permission check before every resolver func so that only administrators can access these methods.

If user is not an admin, they will see an error on the console, or an infinitely loading web page (they should not be able to access the app UI unless they are an admin, so this is not an issue)

Screen Shot 2022-06-29 at 12 38 05 pm

Screen Shot 2022-06-29 at 12 37 59 pm

### CODE REVIEW

File: app/src/resolvers.ts
Comment (mboudreau) [inline app/src/resolvers.ts:30]: can the resolver do multiple functions or only one?  If they can do multiple, just add `checkPermissions` before the inline function.  Another acceptable pattern is to wrap the inline function with the `checkPermissions` which in turn triggers the callback.

File: app/src/resolvers.ts
Comment (mboudreau) [inline app/src/resolvers.ts:15]: `checkPermissions` is not very explicit in terms of naming. It's not a general check for permissions, it's specifically to check if the user is an admin.

File: app/src/resolvers.ts
Comment (mboudreau) [reply app/src/resolvers.ts:30]: I also feel that if forge can't handle authorization, there's a bigger issue here...

File: app/src/check-permissions.ts
Comment (mboudreau) [inline app/src/check-permissions.ts:11]: wouldn't the `asApp()` mean we're calling the API as the app itself which has admin access (which was granted to be installed), hence this check will always return true?  I thought this was specifically a check for the user to be an admin?

**Comment by mboudreau** [comment]
I feel like we should ask the forge team about admin permissions. I feel like this is something forge should be able to handle out of the box and if not, that's a glaringly missing feature.

Also, I'd recommend we also implement [displayConditions](https://developer.atlassian.com/platform/forge/manifest-reference/display-conditions/) so that it doesn't show anything to the user if they're not an admin in the first place so they never get a blank page when it fails :)

File: app/src/resolvers.ts
Comment (yvonnevictoria) [reply app/src/resolvers.ts:15]: Yeah that's fair. I was unsure what to name it as the function I import from the other file is `isAdmin`. Let me think about how to make this clearer.

File: app/src/check-permissions.ts
Comment (yvonnevictoria) [reply app/src/check-permissions.ts:11]: > I feel like we should ask the forge team about admin permissions. I feel like this is something forge should be able to handle out of the box and if not, that's a glaringly missing feature.
> 
> Also, I'd recommend we also implement [displayConditions](https://developer.atlassian.com/platform/forge/manifest-reference/display-conditions/) so that it doesn't show anything to the user if they're not an admin in the first place so they never get a blank page when it fails :)

Currently, if the user is not an admin, they will be unable to access the app anyway. The changes in this PR are to prevent users from calling these functions directly. Assuming the check would happen with `asApp` was the mistake we made initially, that has now been picked up by our security check :)

File: app/src/resolvers.ts
Comment (yvonnevictoria) [reply app/src/resolvers.ts:30]: That's a good point. I can fail from within the function to remove the `checkPermissions` func. However, I was concerned it may not be clear what it returns (error) if it doesn't go through, which is why I opted to have it at the top of the file. Let's discuss on Monday to see if we can come up with a readable option for this..

Comment (yvonnevictoria) [reply app/src/resolvers.ts]: Resolver can only have one func and doesn't support multiple callbacks. 

Discussed on zoom the advantages of this format (wrapping the func) but decided the readability was poor and stuck to the current format
```
resolver.define('getJenkinsServerWithSecret', adminPermissionCheck(async (req) => {
	const jenkinsServerUuid = req.payload.uuid as string;
	return getJenkinsServerWithSecret(jenkinsServerUuid);
}));
```

File: app/src/resolvers.ts
Comment (yvonnevictoria) [reply app/src/resolvers.ts]: Updated name to `adminPermissionCheck` as discussed

File: app/src/check-permissions.ts
Comment (yvonnevictoria) [reply app/src/check-permissions.ts]: Discussed over zoom the results of the security review and how we knew that Forge hasn't included that security check in `asApp`. All discussed that this route should be `asUser`. Have updated this accordingly.

File: app/src/check-permissions.ts
Comment (thombergs) [inline app/src/check-permissions.ts:30]: This error message is the exact same wording as the error message that Jira returns on a 403, if I remember correctly from our previous discussion. This means that if anyone sees this message in the logs, they won't know whether it comes from Jira or from our app :). I would suggest to rephrase this to make clear that the error is coming from the app: "Only Jira administrators can access the Jenkins for Jira admin page".

Comment (yvonnevictoria) [reply app/src/check-permissions.ts:30]: Yep, I had it like that intentionally initially so we didn't need to worry about getting copy approval. Happy to update it :)

**Review by thombergs — APPROVED** [review]

**Review by mboudreau — APPROVED** [review]

**Pull request merged**
